### PR TITLE
Show error dialog for missing runtime/framework

### PR DESCRIPTION
**Pull request by elinor-fung**
On Windows for GUI applications, show error dialog on missing runtime / framework.
- use the default (old, unstyled) message box
- bringback the environment variable for disabling GUI errors - not sure if it is entirely necessary if we aren't showing the dialog for any error, but I remember the dialog causing stress for tools
- add test for error case with GUI apphost

Missing hostfxr (i.e. runtime):
image
Example URL when hitting 'yes': https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=win10-x64

Missing framework:
image
Example URL when hitting 'yes': https://aka.ms/dotnet-core-applaunch?framework=Microsoft.WindowsDesktop.App&version=3.0.0&arch=x64&rid=win10-x64

This PR is targeting 5.0. We do want to port this to 3.1 afterwards.

### CODE REVIEW

**Comment by richlander** [comment]
Cool. It's fun to be doing this all over again. I worked on the .NET Framework version of this. There are three things we need to do:

- Include the minimal amount of information (what is happening, and any currency like a version number or error code)
- Use a link w/the relevant data in query string so that we can take the user to a specific and useful page
- Ask the user if they want to act on the dialog ("Do you want to download .NET Core?" or "Do you want help?") and offers buttons to make that choice, one of which will open the browser. About 1000000% more people will take the call to action if we enable them to launch the browser w/o any effort.

See the original dialog. It largely aligns with these goals.

image

For the first dialog, it should say:

"A fatal error occurred. The required library hostfxr.dll could not be found.

Would you like help with this problem?"

For the second dialog, it says the same thing three different ways. I like the wording of the .NET Framework dialog better. We should write something like:

"To run this application, you must install the following version of .NET Core:
'Microsoft.NETCore.App', version='3.0.0'

Would you like to download this version?"

I know you said that these are just error message, but they are not great. We need simpler messages of these error boxes and better UX with the browser.

**Comment by mairaw** [comment]
I prefer a cleaner and straightforward error message like @richlander is suggesting.
We should definitely review the other dialogs too. Their message is a bit confusing.

**Comment by dagood** [comment]
This comment is probably more specific than we need right now, but I want to make sure it comes up. I think this specific message would make some users install insecure runtimes if they pay too much attention:

> To run this application, you must install the following version of .NET Core:
'Microsoft.NETCore.App', version='3.0.0'

The 3.0.0 there isn't necessarily the latest patch, because FDE apps don't need to be published for each release. It might even say 3.0.0 no matter which 3.0 version you build with, I don't know for sure.

(To be clear about the concern: even with "Yes" opening the correct 3.0 download page, people might see that e.g. 3.0.2 isn't 3.0.0 and manually go find the insecure 3.0.0 to install.)

Something like this maybe:

> To run this application, you must install a version of .NET Core compatible with:
'Microsoft.NETCore.App', version='3.0.0'

**Comment by rowanmiller** [comment]
I think the audience for this message is likely to be non-technical right? Could we display something more like.

> To run this application you need to install .NET Core 3.0 runtime for desktop applications.

I don't think we need to include the complexity of latest patch etc. We can just take folks to the latest patch for install.

We are adding a page to the website, which will give folks a place to get details on exactly what they need to install. I'll be working on it later this week, but I think the URL would be something like https://dotnet.microsoft.com/download/dotnet-core/runtime#desktop

**Comment by richlander** [comment]
I agree with Rowan. We don't need the message to be technically correct but simple and instructive.

**Comment by elinor-fung** [comment]
One thing to note is that if you don't have the runtime installed at all, it currently comes through as the first message about hostfxr (since we look for that first), so perhaps that is the one that should be the most generic and non-technical. Unfortunately, at that point, we don't have specific version or framework information (since the hostfxr library is currently what resolves all that). Maybe that one could just be:
```
To run this application, you must install the .NET Core runtime.
 
Would you like to download the .NET Core runtime?
```

The second error occurs when some specific framework is either missing or an incompatible version (sorry, the screenshot I chose may have been a bit misleading). For example:
- only the runtime is installed not the WindowsDesktop framework - would currently list the missing framework as:
    ```
    The specified framework 'Microsoft.WindowsDesktop.App', version '3.0.0' was not found.
      - No frameworks were found.
    ```
- app requires a version that is incompatible with the one installed (e.g. 3.1 vs 3.0) - would currently list the missing framework as:
    ```
    The specified framework 'Microsoft.WindowsDesktop.App', version '3.1.0' was not found.
      - The following frameworks were found:
          3.0.0 at [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    ```

**Comment by richlander** [comment]
For the first one, how about:

```console
To run this application, you must install .NET Core.

The framework 'Microsoft.WindowsDesktop.App', version '3.1.0' was not found.
 
Would you like to download it now?
```
The problem with the second one is that it is half diagnostics information. These dialogs should target end users. Ideally, the more diagnostic information would be printed elsewhere. For example, if a desktop app is started from cmd.exe, can we just print the diagnostic info to the console? 

This is text that I'd like to see for the second case:

```
To run this application, you must install .NET Core.
 
Would you like to download it now?


```

**Comment by richlander** [comment]
For each of these dialogs, we should define a developer and end-user workflow. For example, most end-users don't care or understand that that they already have a given version of the framework but that it is the wrong one. So, yes, we should drop as much information as possible. They need the quickest path to getting their machine correctly configured and we don't want to confuse them along the way.

For example, given a missing hostfxr.dll, I assume we'd recommend that they install the latest runtime and then go from there? As long as we have a query string that uniquely describes this scenarios, we can decide the policy on the server.

That pattern for collecting error information isn't ideal, but I guess workable. What about: `program.exe 2>&1`?

We need to make this super easy for end-users and then possible for developers.

**Comment by vitek-karas** [comment]
Developer scenario: On Windows we write all the errors into event log as well (and to stderr). Capturing stderr for UI apps is not exactly nice as noted above, but event log should work pretty well (mainly because it's recorded, so if user hits this and then developer comes along, the developer can look at the history to see what the failure was).

The first case - missing `hostfxr` - for 5.0 we should consider baking some kind of special string into the apphost which would provide more information - like the exact version of the runtime, or something similar. Maybe the future TFM (`net5`, net5ui`, ...) - which should be sort of enough for the server to figure out which download to offer. Doing this for 3.1 is probably possible but obviously more risky.

If we rely on the dialog redirecting to a web page, then I think it's probably OK to keep the message int he dialog sort of vague (and simple) - and rely on the web page to offer the right download. This obviously means we need to be able to encode all the necessary information into a query string.

**Comment by elinor-fung** [comment]
Modified the messages per @richlander's latest suggestion and stuck some information in the URL. This is what it looks like now.

Missing hostfxr (runtime)
image
where hitting yes will open https://aka.ms/dotnet-download?runtime=latest

Missing framework
image
where hitting yes will open https://aka.ms/dotnet-download?framework=Microsoft.WindowsDesktop.App&version=3.1.0

File: src/corehost/cli/apphost/apphost.windows.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+        pal::string_t gui_errors_disabled;
+        if (pal::getenv(_X("DOTNET_DISABLE_GUI_ERRORS"), &gui_errors_disabled) && pal::xtoi(gui_errors_disabled.c_str()) == 1)
+            return;
+
+        pal::string_t dialogMsg = _X("To run this application, you must install .NET Core.\n\n");
+        pal::string_t url = DOTNET_CORE_DOWNLOAD_URL;
+        if (error_code == StatusCode::CoreHostLibMissingFailure)
+        {
+            url.append(_X("?runtime=latest"));
```
Comment (vitek-karas) [inline src/corehost/cli/apphost/apphost.windows.cpp]: I think this should be more specific - something like `?noruntime`. I would leave it up to the server to decide which version to get (whether latest or something else).

In 5.0 we will have to stamp the apphost with some version information. I would actually add both the TFM as well as the detailed version of apphost itself (which currently is not stored anywhere in the binary, unfortunately, which will make debugging crash dumps quite a bit painful). I created #8525 for this.

**Comment by rowanmiller** [comment]
Can we also capture the x86 vs x64 in the URL parameters? That would help us give the customer exactly which download they need.

We just published the page that customers will end up on. @richlander and I are going to work on the business logic that needs to sit in the middle and redirect customers to the best version for the app they are trying to run.

Here is an example of the final page customers will see https://dotnet.microsoft.com/download/dotnet-core/3.0/runtime/desktop/x64. The architecture at the end is optional, customers will see both options if none is specified.

**Comment by vitek-karas** [comment]
@rowanmiller That's actually a great point. Ideally we should capture the current RID - not just bitness but platform/OS as well and add that as another query string parameter.

Ideally we want this to work for all platforms, not just Windows - at least eventually.

**Comment by rowanmiller** [comment]
Yep the more info we can get, the better. That way if we need to adjust the content we display in the future, we already have the info we need and it's just website changes.

**Comment by richlander** [comment]
The dialogs look good. Thanks!

Is the aka.ms/dotnet-download link new? We should make sure we have a unique link for this use case so that we curate the experience for it and not worry about other experiences.

**Comment by richlander** [comment]
I agree with @vitek-karas on RID. We (sooner rather than later) will support x86, x64 and arm64 on Windows. It's important that we give the user the right one.

**Comment by elinor-fung** [comment]
@richlander - no, the aka.ms/dotnet-download is what we already had and pointed to for any downloads.
@rowanmiller - would you be the one to ask about how we can get a new one and what it should be?

**Comment by richlander** [comment]
How about: https://aka.ms/dotnet-core-applaunch

File: src/corehost/cli/apphost/apphost.windows.cpp (hunk 1 above)
Comment (elinor-fung) [reply src/corehost/cli/apphost/apphost.windows.cpp]: Changed to `missing_runtime=true`

**Comment by elinor-fung** [comment]
Switched to https://aka.ms/dotnet-core-applaunch. PR description is updated with the current change's behaviour

**Comment by rowanmiller** [comment]
@elinor-fung could you share examples of all the URL patterns that can be generated by this code? We are getting ready to build the service that will receive them and send folks to the best download page.

**Comment by elinor-fung** [comment]
Missing runtime - always has `missing_runtime`, `arch`, and `rid`:
```
missing_runtime=true
arch=<architecture>
rid=<rid>
```
`<architecture>` : architecture of the running app (e.g. x64, x86)
`<rid>`: runtime ID for the current OS and architecture (e.g. win10-x64, win7-x64)

Example:
`?missing_runtime=true&arch=x64&rid=win10-x64`

Missing a specific framework - may not contain the `version`, always has `framework`, `arch`, and `rid`:
```
framework=<frameworkName>
version=<frameworkVersion>
arch=<architecture>
rid=<rid>
```
`<frameworkName>`: name of the missing framework (e.g. Microsoft.NETCore.App, Microsoft.WindowsDesktop.App)
`<frameworkVersion>`: version of the missing framework if known. This follows semantic versioning format of `<major>.<minor>.<patch>-<prereleaseLabel>` (e.g. 3.0.0, 3.0.0-rc1-19456-20)
`<architecture>` / `<rid>`: same as for missing runtime case above

Example:
`?framework=Microsoft.NETCore.App&version=3.0.0&arch=x64&rid=win10-x64`
`?framework=Microsoft.WindowsDesktop.App&arch=x64&rid=win10-x64`

**Comment by elinor-fung** [comment]
@rowanmiller some slight changes to the URL:
- `version` renamed to `framework_version`
- `apphost_version` optionally included

Edited my previous comment to reflect the latest changes.

**Comment by richlander** [comment]
Is there a reason for `missing-runtime`? It seems like that will be the primary case?

**Comment by elinor-fung** [comment]
I had just put that in as an explicit differentiation between there is no runtime installed (and we should probably just point the user to the latest .NET Core runtime) versus there is a runtime installed, but not the frameworks/versions required by the application. If people think that would just be noise, I have no problem with removing it.

**Comment by richlander** [comment]
Ah, if the case for that is "no runtime at all", then I get it. I wanted to avoid the case where we added that param all the time.

It seems like there are three cases:

- Missing specific runtime version.
- Missing a global .NET Core installation.
- Parts of an installation are there, but something is busted and I'm not sure.

I'm not sure if .NET Core has this last case, but .NET Framework did/does.

**Pull request merged**